sql: add user-defined composite types

jordanlewis opened the pull request:
Closes #27792

This commit adds user-defined composite types, which are created with
the `CREATE TYPE t AS (field1 type1, field2 type2, ...)` syntax.

See documentation here:
https://www.postgresql.org/docs/current/rowtypes.html

It permits this kind of thing:

```
CREATE TYPE t AS (a INT, b INT);
CREATE TABLE a (a t);
INSERT INTO a VALUES((1, 2))
SELECT (a).b FROM a
```

Release note (sql change): add user-defined composite column types

jordanlewis committed:
sql: add user-defined composite types

This commit adds user-defined composite types, which are created with
the `CREATE TYPE t AS (field1 type1, field2 type2, ...)` syntax.

See documentation here:
https://www.postgresql.org/docs/current/rowtypes.html

It permits this kind of thing:

```
CREATE TYPE t AS (a INT, b INT);
CREATE TABLE a (a t);
INSERT INTO a VALUES((1, 2))
SELECT (a).b FROM a
```

Release note (sql change): add user-defined composite column types

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/90491)
<!-- Reviewable:end -->

rafiss commented:
super cool! 

two idle comments before a full review:
- let's make sure to add a new cluster version and version gate this.
- do you know how exactly these differ from the implicit table record types in postgres? i was wondering if we could implement them in the exact same way we've already implemented implicit table types, with the addition of marking the table descriptor as unusable. maybe there would be an issue with privileges?

jordanlewis commented:
> let's make sure to add a new cluster version and version gate this.

Definitely.

> do you know how exactly these differ from the implicit table record types in postgres?

There are a couple of differences, but they're pretty minor. It's telling that in Postgres, a UDT appears in `pg_class`. It does strike me that maybe we could do it by just creating a table that you can't insert to, and proxying all the alter statements into alter table statements, but I'm not sure it's a good idea - it seems like it might make existing code more complicated rather than adding new code that's separate.

I'm curious what @ajwerner thinks as the majority of the code here is in schema world but my instinct is that we should model it as a new type kind just like enum.

ajwerner commented:
> I'm curious what @ajwerner thinks as the majority of the code here is in schema world but my instinct is that we should model it as a new type kind just like enum.

That matches my expectation. I can imagine a future migration to add such types for tables too.

jordanlewis commented:
I'm feeling relatively good about contents of this PR, if someone has a chance to review I'd appreciate it. It was nice to learn a little bit about the new schema changer - once you get the hang of it, it seems to do an impressive job of abstracting away the underlying complexity. Or, perhaps I just had blinders on :)

I'm not sure whether I did the dependency tracking right, and I'm sure it's missing a bunch of other stuff as well.

jordanlewis commented:
Also the PR is fairly large, say the word and I'll try to chunk it up a bit.

jordanlewis commented:
Actually, I don't think I have done anything about backup and restore in this patch, so it is probably not complete :)

chengxiong-ruan commented:
Nice work. Maybe it's just me but I found myself had to do quite a bit of context switching while going through the PR briefly. I'd appreciate if you could at least split the PR into several commits : 1. front-end change (parser), 2. legacy schema changer support, 3. fix hydration logic, 4. declarative schema changer support (this worth a separate PR, I think).

jordanlewis commented:
Thanks @chengxiong-ruan I will split it up. Question though, how do I avoid seeing failing tests on the intermediate commits? In particular maybe I need to find a way to tell the declarative schema changer to skip handling DROP TYPE in the case of composite types? I usually like each commit to pass all the tests but maybe that's not a realistic hope at this point.

chengxiong-ruan commented:
> Thanks @chengxiong-ruan I will split it up. Question though, how do I avoid seeing failing tests on the intermediate commits? In particular maybe I need to find a way to tell the declarative schema changer to skip handling DROP TYPE in the case of composite types? I usually like each commit to pass all the tests but maybe that's not a realistic hope at this point.

Ah, yeah, making CI passed on each commit is hard heh. I think in this case for the declarative schema changer, you could raise an unimplemented error [like this](http://github.com/cockroachdb/cockroach/blob/master/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_type.go#L28-L28) so that it can fall back to the legacy schema changer.

jordanlewis committed:
parser: add support for user-defined composite types

Release note: None

jordanlewis commented:
@chengxiong-ruan I have removed the declarative schema changer implementation from this PR and split into 2 commits, one for the parser and one for the rest. I wasn't sure what you meant by fixing the hydration so please let me know if this is helpful, or if you have additional split requests.

chengxiong-ruan commented on pkg/sql/parser/sql.y:
```diff
@@ -9633,6 +9653,26 @@ enum_val_list:
   }
 
+composite_type_list:
+  name simple_typename
```
nit: I think the convention is to wrap `name simple_typename` as a separate object as well. But it looks good to me either way. Feel free to ignore.

jordanlewis replied:
I see a bunch of other places where we do just this so I'll leave it

chengxiong-ruan commented on pkg/sql/catalog/typedesc/type_desc.go:
```diff
@@ -545,6 +545,14 @@ func (desc *immutable) ValidateSelf(vea catalog.ValidationErrorAccumulator) {
+		if desc.Composite == nil {
+			vea.Report(errors.AssertionFailedf("COMPOSITE type desc has nil composite type"))
+		}
+		if len(desc.Composite.TupleLabels) != len(desc.Composite.TupleContents) {
```
I understand that we already defend on frontend that we should always have at least one tuple content, but still worth add extra validation to make sure the lengths are greater than zero.

chengxiong-ruan reviewed (approved):
LGTM!

postamar commented on pkg/sql/catalog/typedesc/type_desc.go:
```diff
@@ -657,6 +669,17 @@ func (desc *immutable) ValidateForwardReferences(
 	}
+
+	if desc.GetKind() == descpb.TypeDescriptor_COMPOSITE {
+		if ids, err := desc.GetIDClosure(); err != nil {
```
This is almost correct. You don't actually want the closure here, only the immediate direct references. It's the established pattern because cross-reference validations can become quite expensive quite quickly.

postamar commented on pkg/sql/catalog/descpb/structured.proto:
```diff
@@ -1440,7 +1442,21 @@ message TypeDescriptor {
+  }
+
+  // composite is the list of fields if this is a composite type.
+  optional Composite composite = 18;
```
Does this serialize correctly if the composite type references enums?

jordanlewis replied:
This is now disallowed.

postamar reviewed (changes requested):
I appreciate the effort but as it is this change is significantly incomplete in terms of testing. I'm not sure there's any way you could have made yourself aware of that. It's unfortunate. In any case I don't have confidence that this new feature actually works, though the implementation seems all quite reasonable, as far as my superficial review is concerned.

You might be able to spare yourself some trouble by forbidding composite types from including other UDTs (including other composite types!) for the time being. Right now, that this seems to be allowed is my main cause of worry as far as the implementation is concerned.

jordanlewis commented on pkg/sql/logictest/testdata/logic_test/composite_types:
```diff
@@ -0,0 +1,113 @@
+DROP TYPE t2
+
+statement ok
+DROP TYPE t
```
Filed #91779 and removed the ability to reference other UDTs from within a composite type.

jordanlewis replied:
Adding this testing has discovered bugs - for example, adding a default expression that includes the user-defined type prevents dropping the user-defined type as expected, but resetting the default expression to something else does not then permit dropping the UDT. There is probably more such stuff. Thanks for recommending it.

jordanlewis replied:
Wait, this is a pre-existing bug in the old schema changer... #91972

jordanlewis commented:
@postamar PTAL, I have:

1. Banned composite types that reference other UDTs
2. Added the indicated missing tests
3. Added commented-out tests for the bug #91972

rafiss reviewed (commented):
does this support arrays of composite types?

rafiss reviewed (changes requested):
can there be tests for arrays of composite types?

postamar commented on pkg/sql/catalog/descpb/structured.proto:
```diff
@@ -1440,7 +1442,21 @@ message TypeDescriptor {
+  }
+
+  // composite is the list of fields if this is a composite type.
+  optional Composite composite = 18;
```
Rather than `Composite` containing slices of metadata which all need to be of the same length, would it be preferable to have `Composite` contain a `repeated CompositeElement` the latter being a sum type for each tuple element?

Right now it doesn't matter very much but at some point we're going to want to support user-defined types and each element is going to have more metadata like an optional forward-reference descriptor ID, etc. So I'm keen for this to evolve gracefully.

postamar commented on pkg/sql/types/types.go:
```diff
@@ -195,8 +195,13 @@ type UserDefinedTypeMetadata struct {
+	// ImplicitRecordType is true if the metadata is for an implicit record type
+	// for a table. Note: this can be deleted if we migrate implicit record types
+	// to ordinary persisted composite types.
+	ImplicitRecordType bool
```
Do we want this to be a table ID instead? Zero meaning it's not an implicit record type, and vice-versa.

jordanlewis replied:
I don't see a use case for recording that table ID, so I guess not? :) We can easily add it if need be.

postamar commented on pkg/sql/schemachanger/scdecomp/decomp.go:
```diff
@@ -189,6 +190,9 @@ func (w *walkCtx) walkType(typ catalog.TypeDescriptor) {
 		}
+	case descpb.TypeDescriptor_COMPOSITE:
+		name := tree.Name(typ.GetName())
+		panic(scerrors.NotImplementedErrorf(&name, "composite types not supported in new schema changer"))
```
Is the schema changer support PR ready for review? I'm just being mindful of accruing technical debt so to be honest I'd rather this change were part of this PR, but won't insist on it. Hopefully removing this `scerrors.NotImplementedErrorf` error should be quite easy at this point.

jordanlewis replied:
Yes, it's ready for review. I moved it out of the PR at @chengxiong-ruan suggestion to keep things small. At this point I do not want to re-add it - can we get this one merged and then I'll work on the other one immediately?

postamar replied:
That's fine by me, thanks! This need to change a zillion things everytime one touches anything in schema-land is sadly a fairly typical feature of our development experience.

postamar commented on pkg/sql/logictest/testdata/logic_test/composite_types:
```diff
@@ -0,0 +1,213 @@
+#
+#statement ok
+#DROP TYPE t;
+#CREATE TYPE t AS (a INT, b TEXT);
```
Let's make a note in #91972 to remember to re-enable this.

jordanlewis replied:
Done.

postamar commented on pkg/sql/logictest/testdata/logic_test/composite_types:
```diff
@@ -0,0 +1,213 @@
+statement error composite types that reference user-defined types not yet supported
+CREATE TYPE t2 AS (t1 t, t2 t)
+
+# Uncomment the below when #90491 is resolved.
```
What's the deal with this? #90491 is this here PR.

jordanlewis replied:
Oops, I meant #91779. Thanks.

postamar commented on pkg/sql/logictest/testdata/logic_test/composite_types:
```diff
@@ -0,0 +1,213 @@
+statement ok
+INSERT INTO tab VALUES ((1, NULL))
+
+# TODO(jordan): this should work fine.
```
Isn't this a rather serious correctness bug? Seems like a showstopper to me at least.

jordanlewis replied:
It's not a correctness bug the way we define it - errors where there shouldn't be errors are not correctness bugs.

The reason this is broken is a type system issue that's unrelated to composite types. I filed #93349 for it but I don't agree that it's a showstopper - it's a classic example of CockroachDB's type system differences from Postgres, and is correctable by adding casts. I've improved the comments here.

Here's an example of the existing broken-ness:

```
demo@127.0.0.1:26257/defaultdb> values((1, 2)), ((1, null));

ERROR: VALUES types tuple{int, unknown} and tuple{int, int} cannot be matched
SQLSTATE: 42804
```

postamar commented on pkg/sql/drop_type.go:
```diff
@@ -249,6 +249,43 @@ func (p *planner) removeBackRefsFromAllTypesInTable(
+			continue
+		}
+		typeIDs = append(typeIDs, id)
+	}
```
nit: consider using `catalog.DescriptorIDSet` here.

postamar replied:
Sorry, I overlooked the fact that `.GetIDClosure()` returned a map and not a slice (or even better, a `catalog.DescriptorIDSet`). My bad!

postamar commented on pkg/sql/create_type.go:
```diff
@@ -370,6 +391,58 @@ func CreateEnumTypeDesc(
+		labels[i] = string(value.Label)
+		typ, err := tree.ResolveType(params.ctx, value.Type, params.p.semaCtx.TypeResolver)
+		if err != nil {
+			return nil, err
```
Should we wrap the error in something which references the label? Seems like things could get confusing otherwise.

postamar replied:
I didn't realize that. Fair enough. Let's stay consistent then.

postamar commented on pkg/sql/create_type.go:
```diff
@@ -222,6 +218,18 @@ func CreateEnumArrayTypeDesc(
+				return nil, unimplemented.NewWithIssue(70099,
+					"cannot use table record type as part of composite type")
+			}
+		}
```
I see these checks when creating the array type descriptor but can't find them when creating the composite-type descriptor, is this expected?

postamar commented on pkg/sql/crdb_internal.go:
```diff
@@ -2633,6 +2636,36 @@ func writeCreateTypeDescRow(
+				ctx, composite.TupleContents[i], resolver.(catalog.TypeDescriptorResolver),
+			); err != nil {
+				return false, err
+			}
```
Just curious: is this hydration logic exercised at all now that we can't reference user-defined types within composite types?

postamar replied:
Agreed in this case.

postamar commented on pkg/sql/catalog/typedesc/type_desc.go:
```diff
@@ -679,6 +688,26 @@ func (desc *immutable) ValidateForwardReferences(
+				vea.Report(errors.AssertionFailedf("referenced type %q (%d) is dropped",
+					typeDesc.GetName(), typeDesc.GetID()))
+			}
+		}
```
Is this code exercised at all now? Shouldn't we simply be erroring when `t.UserDefined()` is true?

jordanlewis reviewed (commented):
Thanks for the review! PTAL.

postamar reviewed (approved):
LGTM. Thanks for doing all of this work!

@rafiss arrays of composite types get exercised in this PR in the new `composite_types` logic test suite, but did you have something else in mind?

jordanlewis commented:
TFTR!

I added the test suite at @rafiss suggestion 😅

jordanlewis commented:
Woohoo! Thanks for the reviews 😁 

bors r+

dvarrazzo commented:
Hello, just stumbled on thia. Psycopg 2 and 3's test suites have an amount of tests involving composite types. Did you try to un-skip them for the crdb version supporting this feature and give them a go? 🙂

jordanlewis commented:
Hey @dvarrazzo, I didn't yet - if I understand our test infrastructure properly, the nightly tests will automatically exercise this tonight and show us which ones are now passing that should be marked as failed. I'll make sure I have the right idea with @rafiss though :-)
